### ISSUE TITLE
Gas limit / Internal JSON-RPC Error when deploying factory pattern contract

### ISSUE THREAD

**Issue by CrackTheCode016** [paritytech/contract-issues#20]
I have tried setting up a Solidity contract using the Factory pattern (i.e., a contract that manages / creates other contracts): 

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.2 <0.9.0;

import './Worker.sol';

contract Factory {
    // Our child contracts
    Worker[] public workers;

    constructor(uint number) {
        workers.push(new Worker(number));
    }

    function addNewWorker(uint number) public {
        workers.push(new Worker(number));
    }

    function getNumberFromWorkerContract(uint index) view public returns(uint) {
        return workers[index].number();
    }
}
```

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.2 <0.9.0;

contract Worker {
    uint public number;
    constructor(uint _number) {
        number = _number;
    }

    function setNewNumber(uint _number) public {
        number = _number;
    }

    function viewNumber() view public returns(uint) {
        return number;
    }
}
```

When I go to deploy the contract, I am presented with the following error: 

```
Internal JSON-RPC error.

You may want to cautiously increase the gas limit if the transaction went out of gas.
```

And before deploying, this modal tells me there might be a problem: 

![Screenshot 2024-12-05 at 8 58 39 AM](https://github.com/user-attachments/assets/67ede28f-6ae7-4176-91c7-bfb23e143294)

This contract works fine on the "normal" Ethereum deployment (RemiX VM).

**Comment by athei** [comment]
This shouldn't work as-is because we don't support deploying new contract code from a contract. Reason is that this is inefficient because the `Worker` code would need to be embedded into the factory code. What this compiles down to on our platform some code that deploys the `Worker` code by hash (it hashes the `Worker` code and embeds the hash into the factory). @xermicus This is how it works, right?

AFAIK it should work when you deploy the `Worker` contract one time manually. Then the code is on-chain and can be referenced by your factory.

But I don't get why it returns an "internal JSON-RPC error". It should return "contract reverted" or something. @pgherveou Do you have an idea?

**Comment by smiasojed** [comment]
@athei @xermicus I needed to deploy the `Worker` contract code first; without it, I encountered `CodeNotFound` errors. After deploying the worker code, I was able to deploy and use the Factory contract successfully.

**Comment by xermicus** [comment]
Yes code hash pre-images must be made available on-chain prior to their instantiation. Support for `EIP-712` should simplify this. Apart from that detail I don't see why the example wouldn't work.

**Comment by smiasojed** [comment]
Does the `deployedBytes` of the `Factory` contract not contain the `Worker` code? The `Factory` contract is very large.

**Comment by xermicus** [comment]
No, it contains the hash image.

**Comment by smiasojed** [comment]
Is it a final solution? Deploying the Worker contract just to be able to create it from the Factory does not make much sense. In the end, you have a spare/useless Worker contract on the chain.

**Comment by xermicus** [comment]
It's only about making the code hash pre-images available. There is no instantiation needed - just uploading the code does it.

**Comment by athei** [comment]
There is no way to do that through the Eth RPC, though. The Eth RPC always bundles code upload and instantiation.

**Comment by xermicus** [comment]
I can't reproduce this in remix. The `Factory` deploys fine and calling `getNumberFromWorkerContract` returns the expected number.

@CrackTheCode016 I therefor close this issue. Feel free to re-open. Thanks for reporting!

**Comment by athei** [comment]
Then most probable cause seems to be a version mismatch between compiler an pallet_revive. @pgherveou Can you still reproduce with latest `master` + latest revive release?

**Comment by pgherveou** [comment]
I might be a few version behind on my current branch will try again shortly

**Comment by pgherveou** [comment]
> Then most probable cause seems to be a version mismatch between compiler an pallet_revive. [@pgherveou](https://github.com/pgherveou) Can you still reproduce with latest `master` + latest revive release?

I tested with the remote compiler 
> Solidity frontend for the revive compiler version 0.1.0-dev.8+commit.fe1b325.llvm-18.1.8

**Comment by athei** [comment]
Probably a spurious error then which is the same as: https://github.com/paritytech/revive/issues/195

Let's continue there.
